html/browsers/browsing-the-web/unloading-documents/beforeunload-canceling.html

```diff
@@ -11,44 +11,90 @@
+  const promise = eventWatcher.wait_for("beforeunload").then(e => {
     assert_true(onbeforeunloadHappened, "CustomEvent must be able to trigger the event handler");
     assert_false(e.defaultPrevented, "The event must not have been canceled");
     window.onbeforeunload = null;
-    t.done();
   });
 
-  window.addEventListener("beforeunload", listener);
-
   window.dispatchEvent(new CustomEvent("beforeunload"));
+
+  return promise;
 }, "Returning a string must not cancel the event: CustomEvent, non-cancelable");
 
-async_test(t => {
+promise_test(t => {
   let onbeforeunloadHappened = false;
   window.onbeforeunload = t.step_func(() => {
     onbeforeunloadHappened = true;
     return "cancel me";
   });
 
-  const listener = t.step_func(e => {
+  const eventWatcher = new EventWatcher(t, window, "beforeunload");
+  const promise = eventWatcher.wait_for("beforeunload").then(e => {
     assert_true(onbeforeunloadHappened, "CustomEvent must be able to trigger the event handler");
     assert_false(e.defaultPrevented, "The event must not have been canceled");
     window.onbeforeunload = null;
     t.done();
   });
 
-  window.addEventListener("beforeunload", listener);
-
   window.dispatchEvent(new CustomEvent("beforeunload", { cancelable: true }));
+
+  return promise;
 }, "Returning a string must not cancel the event: CustomEvent, cancelable");
 
+promise_test(t => {
+  let onbeforeunloadHappened = false;
+  window.onbeforeunload = t.step_func(() => {
+    onbeforeunloadHappened = true;
+    return false;
+  });
+
+  const eventWatcher = new EventWatcher(t, window, "beforeunload");
+  const promise = eventWatcher.wait_for("beforeunload").then(e => {
+    assert_true(onbeforeunloadHappened, "CustomEvent must be able to trigger the event handler");
+    assert_false(e.defaultPrevented, "The event must not have been canceled");
+    window.onbeforeunload = null;
+    t.done();
+  });
+
+  window.dispatchEvent(new CustomEvent("beforeunload", { cancelable: true }));
+
+  return promise;
+}, "Returning false must cancel the event, because it's coerced to the DOMString \"false\" which does not cancel " +
+   "CustomEvents: CustomEvent, cancelable");
+
+// This test can be removed if we update the DOM Standard to disallow createEvent("beforeunload"). Most browsers do not
+// implement that. https://github.com/whatwg/dom/issues/362
+promise_test(t => {
+  let onbeforeunloadHappened = false;
+  window.onbeforeunload = t.step_func(() => {
+    onbeforeunloadHappened = true;
+    return "cancel me";
+  });
+
+  const eventWatcher = new EventWatcher(t, window, "beforeunload");
+  const promise = eventWatcher.wait_for("beforeunload").then(e => {
+    assert_true(onbeforeunloadHappened, "The event handler must have been triggered");
+    assert_false(e.defaultPrevented, "The event must not have been canceled");
+    window.onbeforeunload = null;
+    t.done();
+  });
+
+  const ev = document.createEvent("beforeunload");
```

Comment:
I think the reason Firefox fails this test is because you need to pass `BeforeUnloadEvent` as argument. `createEvent` takes the interface name (and some variants).